File: src/main/java/com/github/highright1234/glacier/protocol/handler/MinecraftDecoder.java
package com.github.highright1234.glacier.protocol.handler;

import com.github.highright1234.glacier.protocol.MinecraftPacket;
import com.github.highright1234.glacier.protocol.BufUtil;
import com.github.highright1234.glacier.protocol.Protocol;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageDecoder;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

@AllArgsConstructor
public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf> {

    @Getter
    @Setter
    private Protocol.Type protocolType;

    private boolean isServer;

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {
        int packetID = BufUtil.readVarInt(msg);
        Protocol.Type.DirectionData dir = isServer ? protocolType.TO_SERVER : protocolType.TO_CLIENT;
        MinecraftPacket packet = dir.getPacket(packetID);
        packet.read(msg);
        if (msg.isReadable()) {
            msg.release();
            return;
        }
        out.add(packet);
    }
}

File: src/main/kotlin/com/github/highright1234/glacier/protocol/handler/MinecraftDecoder.kt
package com.github.highright1234.glacier.protocol.handler

import kotlin.Throws
import java.lang.Exception
import io.netty.buffer.ByteBuf
import com.github.highright1234.glacier.protocol.BufUtil
import com.github.highright1234.glacier.protocol.PipelineUtil
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageDecoder
import com.github.highright1234.glacier.protocol.Protocol

data class MinecraftDecoder(var protocolType : Protocol.Type.DirectionData, val isServer : Boolean) : MessageToMessageDecoder<ByteBuf>() {
    @Throws(Exception::class)
    override fun decode(ctx: ChannelHandlerContext, msg: ByteBuf, out: MutableList<Any>) {
        val packetID: Int = BufUtil.readVarInt(msg).value
        val packet = protocolType.getPacket(packetID)
        packet!!.read(msg, if (isServer)
            PipelineUtil.getClientConnection(ctx.channel())!!.protocolVersion
            else PipelineUtil.getGlacierClient(ctx.channel()!!)!!.version
        )
        if (msg.isReadable) {
            msg.release()
            return
        }
        out.add(packet)
    }
}